sklearn/model_selection/_search.py

```diff
@@ -730,8 +730,10 @@ def evaluate_candidates(candidate_params):
             self.best_params_ = results["params"][self.best_index_]
 
         if self.refit:
-            self.best_estimator_ = clone(base_estimator).set_params(
-                **self.best_params_)
+            # we clone again after setting params in case some
+            # of the params are estimators as well.
+            self.best_estimator_ = clone(clone(base_estimator).set_params(
```

Comment:
Is the following good enough?

```py
clone(base_estimator.set_params(**self.best_params_))
```

Reply:
I don't think so?
We're changing ``base_estimator`` then, right?

Reply:
The `test_grid_search_pipeline_steps` test passes without the double clone. From https://github.com/scikit-learn/scikit-learn/blob/86aea99159c944fae1f519ac61c8b8c484449130/sklearn/model_selection/_search.py#L655 it is already a clone?

Reply:
We could indeed only clone once as @thomasjpfan suggested since `base_estimator` is just a local variable, which isn't used later.

I guess cloning twice is fine too: no surprises.